satellite/satellitedb/overlaycache.go

```diff
@@ -987,6 +987,113 @@ func (cache *overlaycache) UpdatePieceCounts(ctx context.Context, pieceCounts ma
 	return Error.Wrap(err)
 }
 
+// GetExitingNodes returns nodes who have initiated a graceful exit.
+func (cache *overlaycache) GetExitingNodes(ctx context.Context) (exitingNodes storj.NodeIDList, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	rows, err := cache.db.Query(cache.db.Rebind(`
+		SELECT id FROM nodes
+		WHERE exit_initiated_at IS NOT NULL
+		AND exit_finished_at IS NULL
+		`),
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err = errs.Combine(err, rows.Close())
+	}()
+
+	for rows.Next() {
+		var id storj.NodeID
+		err = rows.Scan(&id)
+		if err != nil {
+			return nil, err
+		}
+		exitingNodes = append(exitingNodes, id)
+	}
+	return exitingNodes, nil
+}
+
+// GetExitingNodesLoopIncomplete returns exiting nodes who haven't completed the metainfo loop iteration.
+func (cache *overlaycache) GetExitingNodesLoopIncomplete(ctx context.Context) (exitingNodes storj.NodeIDList, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	rows, err := cache.db.Query(cache.db.Rebind(`
+		SELECT id FROM nodes
+		WHERE exit_initiated_at IS NOT NULL
+		AND exit_loop_completed_at IS NULL
+		AND exit_finished_at IS NULL
+		`),
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err = errs.Combine(err, rows.Close())
+	}()
+
+	for rows.Next() {
+		var id storj.NodeID
+		err = rows.Scan(&id)
+		if err != nil {
+			return nil, err
+		}
+		exitingNodes = append(exitingNodes, id)
+	}
+	return exitingNodes, nil
+}
+
+// UpdateExitStatus returns nodes in exiting status.
+func (cache *overlaycache) UpdateExitStatus(ctx context.Context, request *overlay.ExitStatusRequest) (stats *overlay.NodeStats, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	nodeID := request.NodeID
+
+	updateFields := populateExitStatusFields(request)
+
+	tx, err := cache.db.Open(ctx)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	dbNode, err := tx.Update_Node_By_Id(ctx, dbx.Node_Id(nodeID.Bytes()), updateFields)
+	if err != nil {
+		return nil, Error.Wrap(errs.Combine(err, tx.Rollback()))
+	}
+
+	// Cleanup containment table too
+	_, err = tx.Delete_PendingAudits_By_NodeId(ctx, dbx.PendingAudits_NodeId(nodeID.Bytes()))
```

Comment:
I'm not up to date on the rules. Should updating exit status always imply that all pending audits should be deleted? Either way, the comment above the function should probably say that (or when) this is going to happen.

Reply:
As of this PR: https://github.com/storj/storj/pull/3115
we should not need to worry about deleting from containment outside of the audit service. The audit service will delete a node from containment if the pointer associated with that node has been modified to remove that node (or if the pointer has been deleted). Therefore, we should not need to worry about containment during graceful exit as long as we are removing nodes from the pointers properly.
I removed the delete line for this reason.